src/Build/Evaluation/ItemSpec.cs

```diff
@@ -6,37 +6,169 @@
+            private readonly ItemSpec<P, I> _containingItemSpec;
+            private Expander<P, I> _expander;
+
+            private IMSBuildGlob _msbuildGlob;
+
+            private List<ReferencedItem> _referencedItems;
+            public ExpressionShredder.ItemExpressionCapture Capture { get; }
+
+            public List<ReferencedItem> ReferencedItems
+            {
+                get
+                {
+                    InitReferencedItemsIfNecessary();
+                    return _referencedItems;
+                }
+            }
+
+            protected override IMSBuildGlob MsBuildGlob
+            {
+                get
+                {
+                    if (InitReferencedItemsIfNecessary() || _msbuildGlob == null)
+                    {
+                        _msbuildGlob = CreateMsBuildGlob();
+                    }
+
+                    return _msbuildGlob;
+                }
+            }
+
+            public ItemExpressionFragment(
+                ExpressionShredder.ItemExpressionCapture capture,
+                string textFragment,
+                ItemSpec<P, I> containingItemSpec,
+                string projectDirectory)
+                : base(textFragment, projectDirectory)
+            {
+                Capture = capture;
+
+                _containingItemSpec = containingItemSpec;
+                _expander = _containingItemSpec.Expander;
+            }
+
+            public override int MatchCount(string itemToMatch)
+            {
+                return ReferencedItems.Count(v => v.ItemAsValueFragment.MatchCount(itemToMatch) > 0);
+            }
+
+            public override bool IsMatch(string itemToMatch)
+            {
+                return ReferencedItems.Any(v => v.ItemAsValueFragment.IsMatch(itemToMatch));
+            }
+
+            public override bool IsMatchOnMetadata(IItem item, IEnumerable<string> metadata, MatchOnMetadataOptions options)
+            {
+                if (options.Equals(MatchOnMetadataOptions.CaseSensitive))
+                {
+                    return ReferencedItems.Any(referencedItem =>
+                        metadata.All(m => item.GetMetadataValue(m).Equals(referencedItem.Item.GetMetadataValue(m)) && !item.GetMetadataValue(m).Equals(string.Empty)));
+                }
+                else if (options.Equals(MatchOnMetadataOptions.CaseInsensitive))
+                {
+                    return ReferencedItems.Any(referencedItem =>
+                        metadata.All(m => item.GetMetadataValue(m).ToUpper().Equals(referencedItem.Item.GetMetadataValue(m).ToUpper()) && !item.GetMetadataValue(m).Equals(string.Empty)));
+                }
+                else
+                {
+                    return ReferencedItems.Any(referencedItem =>
+                        metadata.All(m => !item.GetMetadataValue(m).Equals(string.Empty) && NormalizePathMetadata(item.GetMetadataValue(m)).Equals(NormalizePathMetadata(referencedItem.Item.GetMetadataValue(m)))));
```

Comment:
I don't know how performance-sensitive this code is, but if it is, it would be better to precompute some of the NormalizePathMetadata. Specifically, this part: `!item.GetMetadataValue(m).Equals(string.Empty) && NormalizePathMetadata(item.GetMetadataValue(m))` doesn't change per changed referencedItem.